File: amic-platform/src/modules/ma/types/marketing_material.ts
export type MarketingDocType = "TM" | "DM" | "IM";

export type MarketingDocStatus = "DRAFT" | "GENERATING" | "READY" | "CONDITIONAL_READY" | "FAILED";

export interface MarketingMaterial {
  id: string;
  transaction_id: string;
  doc_type: MarketingDocType;
  title: string;
  project_code: string | null;
  status: MarketingDocStatus;
  error_message: string | null;
  source_mode: "GENERATED" | "UPLOADED";
  attachment_id: string | null;
  parameters: Record<string, unknown> | null;
  file_path: string | null;
  file_name: string | null;
  file_size_bytes: number | null;
  quality_score: number | null;
  quality_status: string | null;
  quality_issues: string[] | null;
  slide_count: number | null;
  pipeline_metrics: Record<string, number> | null;
  distribution_eligible?: boolean;
  distributed_to: string[] | null;
  distributed_at: string | null;
  created_by_email: string | null;
  created_at: string;
  updated_at: string;
}

export interface MarketingMaterialSourceRoutingDocument {
  document_id: string;
  original_name: string;
  folder_category: string;
  ddrl_sections: string[];
  primary_workstream: string;
  workstream_tags: string[];
  confidence: number;
  requires_manual_review: boolean;
  reasons: string[];
  is_override?: boolean;
  override_note?: string | null;
  reviewed_by_email?: string | null;
  reviewed_at?: string | null;
  include_for_marketing_material?: boolean;
}

export interface MarketingMaterialSourceRouting {
  version: string;
  summary: {
    total_documents: number;
    included_for_marketing_material?: number;
    excluded_from_marketing_material?: number;
    manual_review_documents: number;
    overridden_documents?: number;
    by_primary_workstream: Record<string, number>;
    target_workstreams?: string[];
  };
  documents: MarketingMaterialSourceRoutingDocument[];
}

export interface MarketingMaterialCreate {
  doc_type: MarketingDocType;
  title: string;
  project_code?: string;
  parameters?: Record<string, unknown>;
  attachment_id?: string;
  distributed_to?: string[];
  distributed_at?: string;
}

export interface UploadedMarketingMaterialInput {
  file: File;
  docType: MarketingDocType;
  title: string;
  projectCode?: string;
  distributedTo?: string[];
  distributedAt?: string;
}

export interface DistributionUpdate {
  distributed_to: string[];
  distributed_at?: string;
}

/** doc_type 한국어 라벨 */
export const MARKETING_DOC_LABELS: Record<MarketingDocType, string> = {
  TM: "Teaser Memo",
  DM: "Discussion Memo",
  IM: "Information Memo",
};

/** 상태 한국어 라벨 */
export const MARKETING_STATUS_LABELS: Record<MarketingDocStatus, string> = {
  DRAFT: "초안",
  GENERATING: "생성 중",
  READY: "완료",
  CONDITIONAL_READY: "조건부 완료",
  FAILED: "실패",
};

/** 품질 상태 라벨 */
export const QUALITY_STATUS_LABELS: Record<string, string> = {
  PASS: "통과",
  CONDITIONAL: "조건부",
  FAIL: "미통과",
  SKIPPED: "미검증",
  LEGACY_UNVERIFIED: "미검증",
};

/** 품질 상태 Badge variant 매핑 */
export const QUALITY_STATUS_VARIANT: Record<string, string> = {
  PASS: "success",
  CONDITIONAL: "warning",
  FAIL: "error",
  SKIPPED: "neutral",
  LEGACY_UNVERIFIED: "neutral",
};

File: deal-mgmt/app/schemas/marketing_material.py
"""마케팅 자료 스키마 — TM / DM / IM."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field, computed_field

from app.models.enums import MarketingDocStatus, MarketingDocType


class MarketingMaterialCreate(BaseModel):
    """마케팅 자료 생성 요청 스키마."""

    doc_type: MarketingDocType
    title: str = Field(..., min_length=1, max_length=300)
    project_code: str | None = Field(
        None, max_length=100, description="프로젝트 코드명 (예: ALPHA). 없으면 transaction에서 자동 생성."
    )
    parameters: dict | None = Field(None, description="memo_generator content JSON. 없으면 기본 템플릿 사용.")
    enable_ralph_loop: bool = Field(False, description="Ralph Loop 품질 강화 모드 활성화")
    ralph_max_iterations: int = Field(3, ge=1, le=10, description="Ralph Loop 최대 반복 횟수")
    ralph_max_cost_usd: float = Field(10.0, ge=1.0, le=50.0, description="Ralph Loop 최대 비용 (USD)")


class DistributionUpdate(BaseModel):
    """배포 대상 목록 업데이트 스키마."""

    distributed_to: list[str] = Field(..., description="배포 대상 회사명 또는 이메일 목록")
    distributed_at: str | None = Field(None, description="배포 일시 (ISO 8601). 없으면 현재 시각.")


class MarketingMaterialOut(BaseModel):
    """마케팅 자료 응답 스키마."""

    id: uuid.UUID
    transaction_id: uuid.UUID
    doc_type: MarketingDocType
    title: str
    project_code: str | None
    status: MarketingDocStatus
    error_message: str | None
    parameters: dict | None
    file_path: str | None
    file_name: str | None
    file_size_bytes: int | None
    quality_score: float | None = None
    quality_status: str | None = None
    quality_issues: list | None = None
    slide_count: int | None = None
    template_fingerprint: str | None = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def pipeline_metrics(self) -> dict[str, int] | None:
        """parameters._metrics에서 성능 메트릭 추출."""
        if self.parameters and isinstance(self.parameters.get("_metrics"), dict):
            return self.parameters["_metrics"]
        return None

    distributed_to: list | None
    distributed_at: str | None
    created_by_email: str | None
    created_at: datetime
    updated_at: datetime

    @computed_field  # type: ignore[prop-decorator]
    @property
    def distribution_eligible(self) -> bool:
        """PASS 상태만 배포 가능."""
        return self.quality_status == "PASS"

    model_config = {"from_attributes": True}


class MarketingMaterialSourceRoutingDocumentOut(BaseModel):
    document_id: uuid.UUID
    original_name: str
    folder_category: str
    ddrl_sections: list[str] = Field(default_factory=list)
    primary_workstream: str
    workstream_tags: list[str] = Field(default_factory=list)
    confidence: float
    requires_manual_review: bool
    reasons: list[str] = Field(default_factory=list)
    is_override: bool = False
    override_note: str | None = None
    reviewed_by_email: str | None = None
    reviewed_at: str | None = None
    include_for_marketing_material: bool = False


class MarketingMaterialSourceRoutingSummaryOut(BaseModel):
    total_documents: int
    included_for_marketing_material: int
    excluded_from_marketing_material: int
    manual_review_documents: int
    overridden_documents: int = 0
    by_primary_workstream: dict[str, int] = Field(default_factory=dict)
    target_workstreams: list[str] = Field(default_factory=list)


class MarketingMaterialSourceRoutingPreviewOut(BaseModel):
    version: str
    summary: MarketingMaterialSourceRoutingSummaryOut
    documents: list[MarketingMaterialSourceRoutingDocumentOut] = Field(default_factory=list)
